### PR TITLE
Enable custom authentication logic

### PR DESCRIPTION
**Pull request by christopher-taormina-zocdoc**
**Description:** 
This PR aims to make creating authentication more flexible. Rather than having to add every possible authenticator to `configauth`, this PR pulls the actual authenticators into extensions that are registered with `configauth` and can looked up by the receiver when being set up.

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector/issues/2101

**Testing:** Tests for the authenticator have not changed, but just moved. Additional tests were added for the new factory class and the configuration.

**Documentation:** There is a new README for the OIDC extension, as well as some expanded information in the `configauth` readme to illustrate how to create an authenticator and register it.

cc @chris-smith-zocdoc @jpkrohling

### CODE REVIEW

File: config/configauth/authenticator.go
Comment (jpkrohling) [inline config/configauth/authenticator.go:83]:
```suggestion
		return fmt.Errorf("authenticator with name %s already registered", name)
```

File: config/configauth/authenticator.go
Comment (jpkrohling) [inline config/configauth/authenticator.go:92]:
```suggestion
		return nil, fmt.Errorf("authenticator not found with name %s", name)
```

File: extension/oidcextension/README.md
Comment (jpkrohling) [inline extension/oidcextension/README.md:3]: Not 100% happy with the module name. I think "oidcauthextension" would be my first option, but "authoidcextension" would keep all auth extensions together. What do you think?

File: extension/oidcextension/README.md
Comment (jpkrohling) [inline extension/oidcextension/README.md:9]: If it has a default, it's not required :-)

File: extension/oidcextension/authenticator.go
Comment (jpkrohling) [inline extension/oidcextension/authenticator.go:52]: Are those changes intentional? Looks like the refactoring was too greedy :-)

File: extension/oidcextension/authenticator.go
Comment (jpkrohling) [inline extension/oidcextension/authenticator.go:49]: Could you leave this change as part of another PR? I'm not sure I agree with this change, especially because our underlying library uses this nomenclature.

File: extension/oidcextension/authenticator.go
Comment (jpkrohling) [inline extension/oidcextension/authenticator.go:134]: Haven't fully checked yet, but do you actually need the host?

File: extension/oidcextension/authenticator.go
Comment (jpkrohling) [inline extension/oidcextension/authenticator.go:157]: Isn't `Start` called by the collector? Why do you need to call it explicitly here?

File: extension/oidcextension/config.go
Comment (jpkrohling) [inline extension/oidcextension/config.go:1]: huh?

File: extension/oidcextension/config.go
Comment (jpkrohling) [inline extension/oidcextension/config.go:43]:
```suggestion
	// The attribute (header name) to look for auth data. Optional, default value: "authorization".
```

File: extension/oidcextension/factory.go
Comment (jpkrohling) [inline extension/oidcextension/factory.go:37]: Does it provide a `WithStart`, where you could specify that your `Start()` function should be called? Same with `WithShutdown`. If it doesn't, your `createExtension` should then call the `Start()` function before adding to the registry.

File: extension/oidcextension/testdata/config.yaml
Comment (jpkrohling) [inline extension/oidcextension/testdata/config.yaml:2]: Could you leave a comment stating that this would fail, because it lacks the required properties?

File: service/internal/resources.go
Comment (jpkrohling) [inline service/internal/resources.go:230]: What are those changes?

**Review by jpkrohling — COMMENTED** [review]
I have a few comments, and one request (to leave an unrelated change for another PR), but I really, really like the direction this is going!

File: config/configauth/README.md
Comment (tigrannajaryan) [inline config/configauth/README.md:5]: Global registries with self-registration are often sources of problems, make testing more difficult, etc. What if I want to have a test with 2 different instances of the extension? (Think of a test which creates 2 full instances of the Collector to test communication between 2 Collectors, which are perhaps configured with different sets of extensions).

A possible alternate is that the caller of createExtension probes the returned value to see if it implements the Authenticator interface and adds it to the Service's local registry of authenticators. We will need to find the way to make this local registry available at the point where the authenticators are created (in  `GRPCServerSettings.ToServerOption` function), not sure what exactly is the best way to do it (possibly add to the ReceiverCreateParams).

Comment (jpkrohling) [reply config/configauth/README.md:5]: > Think of a test which creates 2 full instances of the Collector to test communication between 2 Collectors, which are perhaps configured with different sets of extensions

To make sure I understood this case: this would be with a custom main, where I would create two [`Application`](https://pkg.go.dev/go.opentelemetry.io/collector/service#New) instances as part of the same binary, right? If so, I have to admit that I haven't thought about this case.

I think @bogdandrutu mentioned that he wanted to work on a view registry for the metrics views as well. For Jaeger, we've been thinking about using something similar to this PR here to allow people to extend Jaeger with their own storage mechanisms, where the storage would be a custom `extension` that people can build and register with an extension/component of ours. 

There's definitely a need for registries :-)

Comment (tigrannajaryan) [reply config/configauth/README.md:5]: >To make sure I understood this case: this would be with a custom main, where I would create two Application instances as part of the same binary, right? If so, I have to admit that I haven't thought about this case.

Yes.

>There's definitely a need for registries :-)

Yes, and registries in global variables are undesirable. See for example how we avoided global factory registry although there can be only one per Application. So whatever data we believe is a single instance per Collector needs to be in the Application, not a global variable.

Comment (jpkrohling) [reply config/configauth/README.md:5]: I'll try to scratch something as part of another PR, unless someone wants to work on it first.

File: extension/oidcextension/README.md
Comment (christopher-taormina-zocdoc) [reply extension/oidcextension/README.md:3]: I like `authoidcexetention` to group things.

File: extension/oidcextension/README.md
Comment (christopher-taormina-zocdoc) [reply extension/oidcextension/README.md:9]: True lol. I just wanted to call out that setting, but its probably not that necessary

File: extension/oidcextension/authenticator.go
Comment (christopher-taormina-zocdoc) [reply extension/oidcextension/authenticator.go:134]: Oh no I don't, I stored it when I first moved it in case I thought I would need it but I didn't. Will remove

File: extension/oidcextension/config.go
Comment (christopher-taormina-zocdoc) [reply extension/oidcextension/config.go:1]: Weird I ran the license tool and got this

File: extension/oidcextension/authenticator.go
Comment (christopher-taormina-zocdoc) [reply extension/oidcextension/authenticator.go:52]: lol I think my refactor change name got a little too excited

File: extension/oidcextension/authenticator.go
Comment (christopher-taormina-zocdoc) [reply extension/oidcextension/authenticator.go:49]: I was mostly put off by the change in nomenclature between this error and the config for the extension, since that looks for an `Audience` field while this error will talk about the `ClientID`

File: extension/oidcextension/authenticator.go
Comment (christopher-taormina-zocdoc) [reply extension/oidcextension/authenticator.go:157]: Oh thats a good point, I overlooked that in the change to an extension

File: service/internal/resources.go
Comment (christopher-taormina-zocdoc) [reply service/internal/resources.go:230]: This happened when testing locally and I assumed it was fine. I'll reset this file

File: extension/oidcextension/factory.go
Comment (christopher-taormina-zocdoc) [reply extension/oidcextension/factory.go]: As you mentioned above, the extension does get started by the collector so adding another call isn't necessary

File: config/configauth/README.md
Comment (christopher-taormina-zocdoc) [reply config/configauth/README.md]: I can take a look at it as a part of this PR. 

I think I see two code paths forward right now, that I want to run past y'all since I haven't spent too much time inside the service code.

1. Have the extensions builder would also have to return this registry of Authenticators back up to the service.Application, which can then be passed into the pipeline setup code for the receivers to have access to.
2. Instead of an explicit registry for Authenticators, we can just pass the service's builtExtensions map to the receivers that can expect to find the authenticator as an extension

Let me know if either of those sound right, or if I'm completely off the mark here.

Comment (jpkrohling) [reply config/configauth/README.md]: > I can take a look at it as a part of this PR.

I'd really rather have it as part of another effort, as we need a proper solution for registries in general.

Comment (christopher-taormina-zocdoc) [reply config/configauth/README.md]: @jpkrohling ok this sounds like it might be a bit more sweeping of a change then what I suggested above

Comment (christopher-taormina-zocdoc) [reply config/configauth/README.md]: Can this PR move forward without this wider change to registries or does it need to wait until then?

**Review by tigrannajaryan — CHANGES_REQUESTED** [review]
I don't think we should move forward. Global registries are a global degradation of the codebase quality. I don't want to do that.
Blocking this for now to prevent merging in the current form. I am willing to be convinced.

**Comment by tigrannajaryan** [comment]
To be clear: I believe custom authentication logic is a good idea, but I don't think the current implementation is acceptable.

**Comment by christopher-taormina-zocdoc** [comment]
@tigrannajaryan I guess what I meant was, can I make a change to this PR that enables custom auth without exposing a global registry that may not fix the wider registry problem, but does move this PR away from that implementation

**Comment by tigrannajaryan** [comment]
> I guess what I meant was, can I make a change to this PR that enables custom auth without exposing a global registry that may not fix the wider registry problem, but does move this PR away from that implementation

@christopher-taormina-zocdoc yes, if you update this PR such that there is no longer a global registry then it works for me. Perhaps you can show a draft implementation first if you prefer.

File: component/receiver.go
Comment (christopher-taormina-zocdoc) [inline component/receiver.go:66]: @jpkrohling @tigrannajaryan This commit removes the global registry and passes the built extensions into the ReceiverCreateParams to be used eventually by the configgrpc code. I can clean this up more, but wanted you to take a look before I keep going.

File: config/configgrpc/configgrpc.go
Comment (jpkrohling) [inline config/configgrpc/configgrpc.go:240]: Can you try something different?

* Create an interface somewhere named, say, "WantsServiceExtensions", defining a `SetServiceExtensions(map[string]component.ServiceExtension` function
* In the bootstrap code for the `application`, where it builds the `config` objects, check whether the config that was just built is a `WantsServiceExtensions`, and if so, call `SetServiceExtensions` with the map of extensions.

This way, we have a dependency injection-like mechanism, abstracting this requirement away from the callers, making it easy to use this config with all receivers without them having to manually wire up the map.

Comment (christopher-taormina-zocdoc) [reply config/configgrpc/configgrpc.go:240]: Oh interesting. I might be looking at the wrong place in code, but won't that be earlier than the extensions are built?

Comment (jpkrohling) [reply config/configgrpc/configgrpc.go:240]: I don't remember how the code looks like, but if they are part of the same func/chain, you can iterate over the config objects after you build the extensions.

Comment (tigrannajaryan) [reply config/configgrpc/configgrpc.go:240]: Here https://github.com/open-telemetry/opentelemetry-collector/blob/7eb262845a4ce73782fa967138e7859af4217d9f/service/service.go#L361 extension and receivers are already built, you can call SetServiceExtensions. I would do it uniformly for all components, not just receivers.

Comment (christopher-taormina-zocdoc) [reply config/configgrpc/configgrpc.go]: Pushed a new commit using this approach, let me know what you think.

File: receiver/otlpreceiver/otlp.go
Comment (jpkrohling) [inline receiver/otlpreceiver/otlp.go:235]: Can't we auto-inject directly into the configgrpc? The goal is to remove this knowledge from individual components.

Comment (christopher-taormina-zocdoc) [reply receiver/otlpreceiver/otlp.go:235]: The issue I ran into with trying to inject into the configgrpc, is that there's no guaranteed place for it on an individual component's config. For example, the [otlp receiver has the GRPC server settings object inside a `Protocols` struct](https://github.com/open-telemetry/opentelemetry-collector/blob/a16c599d26b6db7a95489568ca890b06380463cc/receiver/otlpreceiver/config.go#L23) on the config, while the [opencensus receiver just has the GRPC settings as top level attributes on its config struct](https://github.com/open-telemetry/opentelemetry-collector/blob/master/receiver/opencensusreceiver/config.go#L27). This led me to writing a recursive reflection function to iterate over the entire config struct which to me feels a bit messy and overcomplicated.

Comment (jpkrohling) [reply receiver/otlpreceiver/otlp.go:235]: Thinking one abstraction further, perhaps the unmarshaller could check whether the struct it just created implements the new interface?

Comment (christopher-taormina-zocdoc) [reply receiver/otlpreceiver/otlp.go:235]: The config yaml is unmarshalled before any extensions are built, which makes sense because the extensions need their own config to be built

Comment (jpkrohling) [reply receiver/otlpreceiver/otlp.go:235]: You can keep a record of the objects that were built and desire extensions. You can iterate this list later and inject the desired requests.

Comment (christopher-taormina-zocdoc) [reply receiver/otlpreceiver/otlp.go:235]: I don't think getting this information at the time the config is unmarshalled is as straightforward as it seems. After looking through the service code a bit, it seems the service seems can be configured to use a different ConfigFactory at runtime, which any config factory would need to know to return this list of configs that want service extensions. 

Additionally it seems like individual components can be configured with a custom unmarshall function, which it doesn't seem we want want to impose required logic into. 

That seems to leave us back in the place where we would need to reflect over the fully loaded config struct to determine if any portion of it wants the service extensions.

Forgive me if I'm missing an obvious place in the codebase that can be leveraged here.

Comment (jpkrohling) [reply receiver/otlpreceiver/otlp.go:235]: You are probably not missing, I haven't looked at this code for a while. I'll take a look in early Jan to see if I have any new ideas.

**Comment by tigrannajaryan** [comment]
@jpkrohling can you please re-review this?

**Comment by christopher-taormina-zocdoc** [comment]
@tigrannajaryan last I spoke to @jpkrohling he would be around after January 10th or so

**Comment by tigrannajaryan** [comment]
> @tigrannajaryan last I spoke to @jpkrohling he would be around after January 10th or so

@christopher-taormina-zocdoc He is the expert on auth stuff, so let's wait for him to review.

**Comment by jpkrohling** [comment]
I'll take a look in the next couple of days.

**Comment by jpkrohling** [comment]
I haven't forgotten about this one, it was just very hard to look into this with the care that it deserves.

**Comment by jpkrohling** [comment]
Sorry, it's been really hard to get some time to work on this, but here's a very raw version of what I had in mind: https://github.com/jpkrohling/opentelemetry-collector/tree/jpkrohling/2229-inject-authenticator

File: receiver/otlpreceiver/otlp.go
Comment (christopher-taormina-zocdoc) [reply receiver/otlpreceiver/otlp.go:235]: @jpkrohling I think the [draft you posted](https://github.com/open-telemetry/opentelemetry-collector/compare/master...jpkrohling:jpkrohling/2229-inject-authenticator#diff-85ee04e67bda04e6159b92426c87fe161c58d4b5550ef2b0fa7a97214309c125R378) currently leads to this same issue, where the built exporter would still need to be responsible for passing the authenticator to the grpc config settings, instead of being able to push it straight into the grpc config.

Comment (jpkrohling) [reply receiver/otlpreceiver/otlp.go:235]: True, but the config model is the key there. I think it's possible to inject the authenticator to the config. I'll take a look tomorrow.

Comment (jpkrohling) [reply receiver/otlpreceiver/otlp.go:235]: The code editor was still open and the idea fresh in my head, so, I went ahead and pushed another commit, injecting the authenticator in the exporter's config. But you are right, something (exporter? its config?) would still need to make the bridge between the gRPC settings and the authenticator. I never thought I would say this, but I miss Java right now, especially CDI :-)

Comment (jpkrohling) [reply receiver/otlpreceiver/otlp.go:235]: In time: we just missed the train, but I think this would be a good topic for the SIG meeting, to see if other folks have other ideas. https://docs.google.com/document/d/1r2JC5MB7GupCE7N32EwGEXs9V_YIsPgoFiLP4VWVMkE

**Comment by christopher-taormina-zocdoc** [comment]
Sorry I've been busy as of late @jpkrohling, was this discussed in a SIG meeting since you last mentioned it?

**Comment by jpkrohling** [comment]
No, sorry, but I do have in my queue an item to experiment with adding registries to the collector, perhaps as part of the `app`.

**Comment by jpkrohling** [comment]
As an status update: we talked about this during the SIG meeting last week. While the other usage of registries that I had in mind was sorted out, this one hasn't yet.

@tigrannajaryan, @bogdandrutu: I would need your help here. How can we unblock this? People need a way to extend the authenticators with their own logic, and looks like we can't get the list of extensions when building configuration objects.

**Comment by tigrannajaryan** [comment]
Bogdan and I briefly discussed the auth in Collector and believe that there need to be changes in how it is handled generally. I ame not able to review this PR right now, but we will try to settle the auth before the GA.
I am going to reopen this and remove the stale label and it is best to keep this PR as draft for now.
I do not have an ETA no when exactly we can review this PR, we are going over GA-related items one by one and auth is on our list (though we don't yet know if we will have time to solve it by GA).

**Pull request closed without merging**
